This project aims to predict the personality type (Extrovert vs Introvert) of a person based on key characteristics. By analysing patterns between different data types, I aim to explore how specific behavioural traits correleate with one's personality type. 

In [1]:
import pandas as pd
df = pd.read_csv('personality_dataset.csv')
df.head(5)

,Time_spent_Alone,Stage_fear,Social_event_attendance,Going_outside,Drained_after_socializing,Friends_circle_size,Post_frequency,Personality
0,4.0,No,4.0,6.0,No,13.0,5.0,Extrovert
1,9.0,Yes,0.0,0.0,Yes,0.0,3.0,Introvert
2,9.0,Yes,1.0,2.0,Yes,5.0,2.0,Introvert
3,0.0,No,6.0,7.0,No,14.0,8.0,Extrovert
4,3.0,No,9.0,4.0,No,8.0,5.0,Extrovert


In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2900 entries, 0 to 2899
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Time_spent_Alone           2837 non-null   float64
 1   Stage_fear                 2827 non-null   object 
 2   Social_event_attendance    2838 non-null   float64
 3   Going_outside              2834 non-null   float64
 4   Drained_after_socializing  2848 non-null   object 
 5   Friends_circle_size        2823 non-null   float64
 6   Post_frequency             2835 non-null   float64
 7   Personality                2900 non-null   object 
dtypes: float64(5), object(3)
memory usage: 181.4+ KB


In [3]:
df.describe()

,Time_spent_Alone,Social_event_attendance,Going_outside,Friends_circle_size,Post_frequency
count,2837.000000,2838.000000,2834.000000,2823.000000,2835.000000
mean,4.505816,3.963354,3.000000,6.268863,3.564727
std,3.479192,2.903827,2.247327,4.289693,2.926582
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2.000000,2.000000,1.000000,3.000000,1.000000
50%,4.000000,3.000000,3.000000,5.000000,3.000000
75%,8.000000,6.000000,5.000000,10.000000,6.000000
max,11.000000,10.000000,7.000000,15.000000,10.000000


In [4]:
df.isnull().sum()

Time_spent_Alone             63
Stage_fear                   73
Social_event_attendance      62
Going_outside                66
Drained_after_socializing    52
Friends_circle_size          77
Post_frequency               65
Personality                   0
dtype: int64

In [5]:
df[['Time_spent_Alone', 'Social_event_attendance','Going_outside', 'Friends_circle_size','Post_frequency']].mean()

Time_spent_Alone           4.505816
Social_event_attendance    3.963354
Going_outside              3.000000
Friends_circle_size        6.268863
Post_frequency             3.564727
dtype: float64

In [6]:
df[['Time_spent_Alone', 'Social_event_attendance','Going_outside', 'Friends_circle_size','Post_frequency']] = df[['Time_spent_Alone', 'Social_event_attendance','Going_outside', 'Friends_circle_size','Post_frequency']].fillna({
    'Time_spent_Alone' : 4.505816 ,
    'Social_event_attendance' : 3.963354,
    'Going_outside' : 3.000000,
    'Friends_circle_size' : 6.268863,
    'Post_frequency' : 3.564727
})
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2900 entries, 0 to 2899
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Time_spent_Alone           2900 non-null   float64
 1   Stage_fear                 2827 non-null   object 
 2   Social_event_attendance    2900 non-null   float64
 3   Going_outside              2900 non-null   float64
 4   Drained_after_socializing  2848 non-null   object 
 5   Friends_circle_size        2900 non-null   float64
 6   Post_frequency             2900 non-null   float64
 7   Personality                2900 non-null   object 
dtypes: float64(5), object(3)
memory usage: 181.4+ KB


In [7]:
df[['Stage_fear', 'Drained_after_socializing']].mode()

,Stage_fear,Drained_after_socializing
0,No,No


In [8]:
df[['Stage_fear', 'Drained_after_socializing']] = df[['Stage_fear', 'Drained_after_socializing']].fillna('No')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2900 entries, 0 to 2899
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Time_spent_Alone           2900 non-null   float64
 1   Stage_fear                 2900 non-null   object 
 2   Social_event_attendance    2900 non-null   float64
 3   Going_outside              2900 non-null   float64
 4   Drained_after_socializing  2900 non-null   object 
 5   Friends_circle_size        2900 non-null   float64
 6   Post_frequency             2900 non-null   float64
 7   Personality                2900 non-null   object 
dtypes: float64(5), object(3)
memory usage: 181.4+ KB


In [9]:
df['Stage_fear'] = df['Stage_fear'].astype(str).str.strip().str.lower()
df['Drained_after_socializing'] = df['Drained_after_socializing'].astype(str).str.strip().str.lower()

In [10]:
df['Stage_fear'] = df['Stage_fear'].map({'no' : 0.0, 'yes' : 1.0})
df['Drained_after_socializing'] = df['Drained_after_socializing'].map({'no' : 0.0, 'yes' : 1.0})
df['Personality'] = df['Personality'].map({
    'Introvert' : 0.0, 
    'Extrovert' : 1.0
})
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2900 entries, 0 to 2899
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Time_spent_Alone           2900 non-null   float64
 1   Stage_fear                 2900 non-null   float64
 2   Social_event_attendance    2900 non-null   float64
 3   Going_outside              2900 non-null   float64
 4   Drained_after_socializing  2900 non-null   float64
 5   Friends_circle_size        2900 non-null   float64
 6   Post_frequency             2900 non-null   float64
 7   Personality                2900 non-null   float64
dtypes: float64(8)
memory usage: 181.4 KB


In [11]:
df['Drained_after_socializing'].head(10)

0    0.0
1    1.0
2    1.0
3    0.0
4    0.0
5    0.0
6    0.0
7    0.0
8    1.0
9    0.0
Name: Drained_after_socializing, dtype: float64

In [12]:
x = df[['Time_spent_Alone', 'Stage_fear', 'Social_event_attendance', 'Going_outside', 'Drained_after_socializing', 'Friends_circle_size', 'Post_frequency']]
y = df['Personality']

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(x,y, test_size = 0.2, random_state = 0)

I split the data using the 80/20 rule instead of the 70/30 rule because the dataset I am working with is large and a better split would be 80/20. 

In [14]:
from sklearn.tree import DecisionTreeClassifier
decision_tree_model = DecisionTreeClassifier(random_state = 0)
decision_tree_model.fit(X_train, Y_train)

DecisionTreeClassifier(random_state=0)

In this the decision tree model was created, while keeping the random_state = 0, to ensure that the result stays the same. 

In [15]:
Y_predic = decision_tree_model.predict(X_test)
Y_predic

array([1., 0., 1., 1., 1., 1., 0., 1., 0., 0., 0., 1., 0., 1., 0., 1., 1.,
       0., 0., 1., 1., 1., 0., 1., 1., 0., 1., 0., 0., 0., 1., 0., 0., 1.,
       1., 1., 1., 1., 1., 0., 1., 1., 0., 0., 1., 0., 1., 1., 1., 0., 0.,
       0., 1., 0., 0., 0., 0., 1., 1., 1., 0., 0., 0., 0., 0., 1., 0., 1.,
       1., 0., 1., 1., 1., 0., 0., 0., 1., 0., 0., 1., 1., 0., 0., 1., 1.,
       0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0., 1., 0., 1., 1., 0., 0.,
       0., 0., 0., 1., 1., 0., 1., 1., 0., 1., 1., 0., 1., 0., 1., 0., 1.,
       1., 1., 0., 1., 1., 0., 1., 1., 0., 1., 0., 0., 0., 1., 0., 0., 0.,
       0., 1., 1., 1., 0., 0., 1., 1., 1., 0., 0., 0., 1., 1., 1., 1., 1.,
       0., 1., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 1., 0., 0., 1., 0.,
       0., 1., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 1., 0.,
       0., 0., 0., 1., 1., 1., 0., 0., 1., 1., 1., 0., 0., 1., 0., 0., 1.,
       1., 0., 0., 1., 0., 1., 0., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0.,
       1., 0., 1., 0., 1.

In [16]:
from sklearn.metrics import confusion_matrix
conmat_pred = confusion_matrix(Y_test, Y_predic)
conmat_pred

array([[273,  37],
       [ 36, 234]])

In [17]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [18]:
accuracy = accuracy_score(Y_test, Y_predic)
accuracy

0.8741379310344828

In [19]:
precision = precision_score(Y_test, Y_predic)
precision

0.8634686346863468

In [20]:
recall = recall_score(Y_test, Y_predic)
recall

0.8666666666666667

In [21]:
f1 = f1_score(Y_test, Y_predic)
f1

0.8650646950092421

Interpretation: The model correctly identified 273 Introverts and 234 Extroverts, while misidentifying 37 introverts as extroverts and 36 extroverts as introverts. In conclusion, out of 580 test samples, 507 were correctly identified, giving the model an accuracy of 87.4%. Apart from this, the model has a precision rate of 86.3%, recall rate of 86.7% and F1 score of 86.5%, illustrating the effective prediction of personality type, with a minimal misclassification rate and fairly equal performance in both classes. 


In [22]:
from sklearn.metrics import classification_report
class_rep = classification_report(Y_test, Y_predic)
class_rep

'              precision    recall  f1-score   support\n\n         0.0       0.88      0.88      0.88       310\n         1.0       0.86      0.87      0.87       270\n\n    accuracy                           0.87       580\n   macro avg       0.87      0.87      0.87       580\nweighted avg       0.87      0.87      0.87       580\n'

Conclusion

After performing comprehensive data preprocessing and feature encoding, a Decision Tree Classifier was developed to categorize personality types into groups. The model demonstrated strong classification performance, achieveing an overall accuracy of 87.4% on the test dataset. 

Results

Final Accuracy Score: 87.4%
F1-Score: 0.865 (86.5%)


Accuracy Score Interpretation

The model achieved an accuracy of 87.4%, correctly classifying 507 individuals (273 Introverts and 234 Extroverts). This high success rate indicates that the model's decision-making logic is highly effective, correctly identifying the personality type for nearly 9 out of every 10 individuals.

F1-Score Interpretation

The model achieved a F1-Score of 86.5%, confirming that the model is robust and unbiased. It demonstrates a successful balance between precision and recall, ensuring that the model is equally accurate and fair both classification types (Introvert and Extrovert). 
